runtime/v2/balena/balena.go

```diff
@@ -0,0 +1,79 @@
+package balena
+
+import (
+	"os"
+	"os/exec"
+	"path"
+
+	"github.com/containerd/containerd/plugin"
+)
+
+// DisableShimPlugins returns true if the plugin represented by r should be
+// disabled.
+//
+// While migrating to the shim runtime v2, I shed many tears to understand this,
+// so I'll try to explain it in detail for the future generations and the future
+// me. I hope I got this right.
+//
+// containerd-shim-runc-v2 seems to be strongly based on a plugin system that is
+// basically the same containerd itself used before. So, at startup, both will
+// load all registered plugins. How are these plugins registered? The plugins
+// register themselves by calling `plugin.Register()` (from
+// `github.com/containerd/containerd/plugin`) in an `init()` function. So far so
+// good.
+//
+// Things start to get tricky when we realize that containerd and
+// containerd-shim-runc-v2 need a different set of plugins. In standard Moby
+// this is not a problem, because each of these binaries will import a different
+// set of packages, and therefore will only run the correct subset of `init()`s,
+// and thus will register only the plugins they need.
+//
+// The problem is that in balenaEngine we have one single binary that
+// amalgamates containerd, containerd-shim-runc-v2, and more. So, that single
+// big binary will run every `init()` from every package imported by everyone.
+// As a result, containerd-shim-runc-v2 will try to load all plugins in the
+// world -- including some that will cause it to crash.
+//
+// (Sidebar: Why do some plugins cause containerd-shim-runc-v2 to crash? AFAIU,
+// it's essentially because the shim plugin system is not as advanced as the one
+// in containerd. For example, containerd-shim-runc-v2 will not pass
+// "configurations" to the plugins, so plugins that expect to get configs will
+// crash. For details, see the `run()` function in
+// `vendor/github.com/containerd/containerd/runtime/v2/shim/shim.go` and search
+// for "TODO". There is some code that would prepare the configs so they could
+// be later accessed by the plugins, but it is commented out since forever. This
+// commented out code seems to come from
+// `vendor/github.com/containerd/containerd/services/server/server.go`, which I
+// believe is the equivalent containerd code.)
+//
+// And this finally brings us to this function. We use it to filter out the
+// containerd-shim-runc-v2 plugins that we don't want to load. Or rather, to
+// filter out all plugins except the few ones that really needed.
+//
+// The End.
+func DisableShimPlugins(r *plugin.Registration) bool {
+	// These are the plugins we want to load.
+	want := (r.Type == plugin.TTRPCPlugin && r.ID == "task") ||
```

Comment:
It looks like we are picking specific plugins here, which means that addition or removal of a plugin requires a code change. Is there any characteristic or property of a plugin that can be used to identify a plugin that should be enabled?

Does the list change often? Would it be useful for the list to be provided in a configuration file?

Reply:
Good points, thanks for pointing them out!

> Is there any characteristic or property of a plugin that can be used to identify a plugin that should be enabled?

These plugins provide core functionality really. As far as I can tell, it's just a matter of how they architected the shim. The only commonality I see is that these are the plugins registered from code located under `runtime/v2/shim/`, but the plugin system doesn't seem to allow to check for this. (Interestingly, I also see no provided way to replace any of these plugins with something else. Maybe at some point they had plans to move more into a "true" plugin-based design, or I am  missing something.)

> Does the list change often? Would it be useful for the list to be provided in a configuration file?

Did a quick check and this list seems to be stable since the plugin architecture was adopted ~2 years ago.